File: 程序员面试金典/1.8.零矩阵.py
from typing import List


class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m = len(matrix)
        if m == 0:
            return

        n = len(matrix[0])

        a = []
        b = []

        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    a.append(i)
                    b.append(j)

        for i in range(m):
            for j in range(n):
                if i in a or j in b:
                    matrix[i][j] = 0

File: 程序员面试金典/1.8.零矩阵.cpp
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix)
    {
        int M = matrix.size();
        int N = matrix[0].size();

        unordered_set<int> hang;
        unordered_set<int> lie;

        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                if (matrix[i][j] == 0) {
                    hang.insert(i);
                    lie.insert(j);
                }
            }
        }

        for (int i = 0; i < M; i++) {
            if (hang.count(i)) {
                for (int j = 0; j < N; j++) {
                    matrix[i][j] = 0;
                }
            } else {
                for (int j : lie) {
                    matrix[i][j] = 0;
                }
            }
        }
    }
};